Fine-grained debugging information in Cmm

mshinwell opened the pull request:
This patch reorganises the provision of debugging information in the Cmm language.  This is important as it yields fine-grained information that should provide good stepping behaviour in the debugger.  This patch also introduces debugging information on `Cswitch` expressions, which a later patch will provide from the middle end.

Apart from the change in debugging information output this patch should be a no-op.  I need to make another pass over it to ensure nothing else has slipped in, since this has been extracted from a branch with a much larger diff.

The changes are entirely mechanical although tedious in the extreme in `Cmmgen`.  I think it is the case that the appearance of the Cmm language is more uniform after this change.  My tired fingers ask that we try to review this promptly so as to minimise the risk of any merge conflicts.

P.S. Note that the `cmmgen.ml` diff isn't shown by Github.

mshinwell committed:
More debuginfo in Cmm

mshinwell committed:
tidying

alainfrisch commented:
Moving the debuginfo for operations to the Cop constructors makes a lot of sense.  The other change to cmm.mli is the addition of a debuginfo argument on Cswitch, but in the current PR, it is always none.  Do you intend to add plumbing later to keep a debuginfo for Cswitch?

nojb commented:
From the PR description:

> This patch also introduces debugging information on Cswitch expressions, which a later patch will provide from the middle end.

mshinwell commented:
@alainfrisch Yes, that's right.  I don't think I can reasonably prepare any of the other patches in this area until this one is in, though.

gasche commented:
My understanding is that previously, `Debuginfo.t` was only added to constructors representing either function calls or operations that could "fail" (`Cextcall`, `Calloc`, `Craise`, `Ccheckbound`). We may call that a "sparse" debug information.

You generalize the representation (making it more uniform) by making debug information available for all operations whatsoever -- "dense" debug information. The rationale, if I understand correctly, is to be able to refer back to the source location on arbitrary program points on which the debugger may be stopped?

I reviewed the code and I checked that it's mostly a mechanic transformation, passing more `dbg` parameters around. I trust it preserves correctness -- or that any remaining issue would be better found by testing rather than human review. I can't comment on the impact on the backend -- sensibly more information is provided now, and more will come as the middle-end is updated. Is the size of debug informations an issue for some people (@alainfrisch?).

## General remarks

The `clambda.mli` representation is unchanged by this patch, and it also uses sparse debug information. So in `cmmgen` many of the cases start with `let dbg = Debuginfo.none in` as there is no actual debug information available. I suppose this is what you mean by "a later patch will provide information for the middle-end" -- the clambda and flambda paths will also be changed to provide more information.

The patch does not try to provide debug information in the switch-generating code (or in `asmcomp/strmatch`). Is that planned for later?

The patch does not try to provide debug information in the arity-manipulating functions (`apply_function{,_body}`, `{send,tuplify,final_curry}_function`, etc.), is that planned for later?

(The patch also made me wonder whether a record for `Cop` wouldn't be beneficial long-term, but I get the point that making such changes is not fun and I would support merging the current type result. If you are going to implement the changes in ulambda/flambda later, would it make sense to consider this question now?)

You made the choice of changing the behavior of `Printcmm`, printing debug information on all operations. Another possible choice would be to have propagated the `dbg` parameter to the `operation` call and keep the previous behavior. I would agree that printing all available information is a reasonable choice, but isn't there a risk of being flooded with too much debug info compared to the rest of the cmm code? What's the impact on the readability of the `-dcmm` output, would you show a small before/after example?

## Minor remarks

In `Cmmgen.transl`, most cases that lack debug information from clambda start with `let dbg = Debuginfo.none`, but this style is not consistently used, `Uassign` and `Uunreachable` inline the definition. Could you change those cases back to the common style? It makes it easier to scan for missing frontend information.

I would have expected `Cmmgen.transl_let{,rec}` to take a `dbg` parameter. Is there a reason for the choice of locally using `Debuginfo.none` instead? Do you know for sure that debug locations would not help there?

``` diff
   | Uclosure(fundecls, clos_vars) ->
       let block_size =
         fundecls_size fundecls + List.length clos_vars in
+      let dbg = ref None in
       let rec transl_fundecls pos = function
           [] ->
             List.map (transl env) clos_vars
         | f :: rem ->
             Queue.add f functions;
+            if !dbg = None then dbg := Some f.dbg;
             let header =
               if pos = 0
               then alloc_closure_header block_size f.dbg
               else alloc_infix_header pos f.dbg in
             if f.arity = 1 || f.arity = 0 then
               header ::
               Cconst_symbol f.label ::
               int_const f.arity ::
               transl_fundecls (pos + 3) rem
             else
               header ::
               Cconst_symbol(curry_function f.arity) ::
               int_const f.arity ::
               Cconst_symbol f.label ::
               transl_fundecls (pos + 4) rem in
-      Cop(Calloc Debuginfo.none, transl_fundecls 0 fundecls)
+      Cop(Calloc, transl_fundecls 0 fundecls,
+        match !dbg with None -> Debuginfo.none | Some dbg -> dbg)
```

Wouldn't it make sense to use `Debuginfo.concat` here?

`exit_if_true` and `exit_if_false` use their debug parameters differently (the former ignores it, the latter uses it). Why?

alainfrisch commented:
Is there a clear criterion to decide which constructors in Cmm.expression should have a Debuginfo.t?  If more are expected, what about turning that type into a record type holding a Debuginfo.t next to a "descr" field (as done in Parsetree and Typedtree)?

mshinwell commented:
@alainfrisch You are probably right that it would be better to put it on every expression node.  I think that would need to be done as a subsequent patch.

chambart commented on asmcomp/cmm.mli:
```diff
@@ -134,10 +134,10 @@ and expression =
   | Csequence of expression * expression
   | Cifthenelse of expression * expression * expression
-  | Cswitch of expression * int array * expression array
+  | Cswitch of expression * int array * expression array * Debuginfo.t
```
Why doesn't Cifthenelse get its debuginfo too ?

mshinwell replied:
It probably should -- see reply to @alainfrisch .  I think it was just that I'd empirically noticed that switches seemed to be worse (perhaps because the scrutinee is often just a variable?).

mshinwell replied:
(The point being that if it's Cifthenelse on a variable, then there is a comparison operator in the first subexpression containing debug info; however for Cswitch, that is absent.)

lpw25 replied:
The idea was to put debuginfo on those constructs which correspond to actual instructions in the output. We decided that mostly meant `Cop`, `Cswitch` and `Cifthenelse`, but then I think Mark decided to skip `Cifthenelse` since it didn't turn out to be particularly important.

I'm not too sure about putting debuginfo on everything. Things like `let` do not really correspond to resulting instructions, or indeed to a specific piece of the source program, so it is not clear that attaching debuginfo to it really makes sense.

alainfrisch replied:
> which correspond to actual instructions in the output

Cassign (resp. Cexit) is likely to result in a "mov" (resp. "jmp") instruction in the result, no?  Cloop will also produce instructions which can be attributed to that specific loop.  Both Cassign and Cloop can also be traced back to a specific piece of the source program (less so for Cexit).

mshinwell replied:
(As an aside on these two, Leo and I did wonder if in the long term we might be able to make Cmm be in SSA form, which it nearly is already, save for constructs like these.  The context was whether it might open the door for better register allocation algorithms.)

lpw25 replied:
I'm a bit torn on those ones. In a sense `mov` and `jump` don't really do anything and so never really map back properly to the source program. The registers and the code layout bare only a passing resemblance to the original code, so the user doesn't care about them. Essentially, I see these instructions as just artifacts of the need to use a finite set of registers and a linear code representation, so for my reasoning about what gets some debug info I think they in some sense "don't count".

alainfrisch replied:
Cassign typically represents an assignment to a local reference.  I'm not sure to see why this would not really do anything nor why this wouldn't map back to the source program.

lpw25 replied:
Well, this probably comes from taking an SSA-style view on things, but to me `Cassign` is just the same as `Clet`: it doesn't actually do anything at runtime, it just tells the compiler to give an existing thing a particular name. To me, when you turn a local reference into a `Cassign` you are removing the mutable thing and replacing it by just passing the most recent value around to the appropriate locations -- the act of mutating has been taken out of the semantics.

If/when we switch to SSA-style that would be more obviously true, rather than just my interpretation of things.

alainfrisch replied:
I probably completely miss the context of the PR, but from a user perspective, being able to put a breakpoint before an assignment, show the current value of the reference, step through the assignment, and show the new value seem like a basic use of a debugger; and this should not be broken if the reference is turned into a local variable by the compiler.  But perhaps this can be achieved already without keeping any more information.

lpw25 replied:
I think that stepping through the mutation may not be possible. This is due to missing features of DWARF -- although there are proposals to improve things in DWARF 6.

Basically, when you optimise something out of a program, DWARF should provide you with a way to artificially put it back in again. For example, it lets you describe values which have been optimised away. However, the stepping behaviour is not directly specifiable in DWARF yet, instead debuggers just runs the program until the line number associated with the current pc changes, so there is no way to tell it about an operation (e.g. the mutation of an optimised local variable) which has been removed. (This is also what prevents expression level stepping in gdb and lldb).

alainfrisch replied:
Thanks for clarifying.  So, in a future were DWARF would be improved in this direction, do you agree it would be useful to keep the debuginfo on Cassign?

lpw25 replied:
Not quite. In that future, I would expect the optimisation of a local reference to produce both `Cassign` and some "phantom" operation like the "phantom" let in Mark's other PR. The phantom version would have the debuginfo and the `Cassign` wouldn't.

Well actually by the time DWARF 6 is out I would hope we had removed `Cassign` in favour of SSA -- but the "phantom" reference mutation would still exist and would still have debuginfo attached.

mshinwell replied:
@alainfrisch We could put information on `Cassign` at the moment, but the problem is that it's not always going to yield anything useful (depending on what register allocation does).  If you feel strongly we can add it.

lpw25 replied:
(Note that I'm not particulary against adding debuginfo to `Cassign`, it will probably help sometimes at the moment -- in the case where the `mov` happens to still be present in the output. But in principle I don't think it should be there, and in the long run I think that it should be removed -- or possibly `Cassign` itself should be removed)

alainfrisch replied:
I don't feel strongly either way.  But providing some guidelines (perhaps as comments in the code) about how to decide where debuginfo are relevant could be helpful.

mshinwell committed:
code review

mshinwell commented:
@gasche Thanks for the review.  The rationale is as you say, but also the converse: if debugging information isn't fine-grained enough, single-stepping in the debugger will jump over blocks of code.

Debugging information on switches will be provided in a subsequent patch.

The various arity-manipulating functions don't need to have any debug information at least for the moment, I don't think.  Users are unlikely to want to step through those.

I don't think it is worth the churn to try to change `Cop` to a record at present.  I also think it would lead to a net decrease in readability for not much gain.

I will check how verbose the `-dcmm` output is.

I have dealt with the `Uassign` and `Uunreachable` cases as you suggest.

For transl_let and transl_letrec, I have now used the `let dbg = Debuginfo.none` idiom.  There is a change in a subsequent patch to alter the type of all Cmm identifiers when they appear in binding position (e.g. `dbg` in `let dbg = ...`) such that they carry more information.  That information could be used to populate the `dbg` variable here.  Although @alainfrisch 's suggestion would be better still.

For the block of code you quote, `Debuginfo.concat` is not the right thing.  That is used to express a stack of inlined-out frames.  However now I look at it I see this code is actually wrong; it should be using the location where the closure was allocated, not the location of the function definitions involved.  I think we can just use `Debuginfo.none` for the moment since the closure should always be bound to a variable that did not occur in the user's source code, and thus should not appear in the debugger.

I have fixed up `exit_if_true` and `exit_if_false`, although these functions undergo a significant refactoring in a subsequent patch, since they currently rely too much on the exact structure of the Cmm expression.

gasche commented:
You have a Continuous Integration failure as a test should be updated:

```
Running tests from 'tests/asmcomp' ...
File "parsecmm.mly", line 23, characters 2-32:
Error: The constructor Cswitch expects 4 argument(s),
       but is applied here to 3 argument(s)
```

Thanks for your replies. I think the question of the impact on `-dcmm` is the only remaining one for me. Is the `print_compact` PR related?

mshinwell committed:
Make Printcmm behave like it used to

mshinwell committed:
Comment

mshinwell committed:
Fix printcmm.mly

mshinwell committed:
Merge remote-tracking branch 'ocaml/trunk' into cmmgen_dbg

mshinwell commented:
Recent updates following comments from @gasche and @alainfrisch :
- Reverted `Printcmm`'s behaviour to what it used to be.  I think it's going to be too verbose otherwise.  We can increase verbosity in the future if needs be (possibly via an extra switch).
- Fixed `printcmm.mly` in the testsuite.
- Added a comment mentioning why not every expression node has a `Debuginfo.t` value.
- Merged with latest trunk.

Are there any other comments?

mshinwell commented:
@gasche The `print_compact` PR is not related to this one.

gasche commented:
Indeed we have `-drawlambda` and `-dlambda`, or `-dparsetree` and `-dsource`, it would make sense to have `-drawcmm` and `-dcmm`.

I don't have any comments left, and as far as I'm concerned this is good for merging. @alainfrisch, what's your opinion?

alainfrisch commented:
Looks good to me.

gasche commented:
Hm, it looks like my `check_all_arches` nets in the CI [just caught some fish](https://travis-ci.org/ocaml/ocaml/jobs/168291252#L4735). Is it an actual build issues on other architectures?

mshinwell commented:
@gasche This isn't the same as `-drawlambda` / `-dlambda` though.  Those are potentially materially different due to the action of `Simplif`.  However I think we may need to introduce a more general "verbose debug" flag in the light of all this debugging work, since the extra information does sometimes need to be seen, but only really for debugging the debugging-related parts.

I'm looking at `check_all_arches` now.

mshinwell committed:
Fix all architectures

mshinwell commented:
I've fixed all architectures.  `make check_all_arches` works for me now.

gasche commented:
I went ahead and merged, thanks!
